Site editor global styles settings: ensure generated global styles are correct in canvas between views

ramonjd opened the pull request:
closes #73350 

## What? How?

Selecting a theme style variation in the Site Editor does not persist between the editor view (`canvas=edit`) and the main design view (`canvas=view`). The styles appear correctly in the editor view but remain stale in the main design view.

This PR fixes that by generating global styles directly from the merged config in `useSpecificEditorSettings()` instead of reading from either store.

## Why? 

A git bisect led me to this PR:

- https://github.com/WordPress/gutenberg/pull/72681

**Before PR #72681:**
- `GlobalStylesRenderer` updated `editSiteStore` via `updateSettings()`
- `useSpecificEditorSettings()` read from `editSiteStore` via `getSettings()`
- Both used the same store, so styles stayed in sync ✅

**After PR #72681:**
- `GlobalStylesRenderer` now updates `editorStore` via `updateEditorSettings()`
- `useSpecificEditorSettings()` still reads from `editSiteStore` via `getSettings()`
- They use different stores, so the main design view doesn't see updates ❌

The entity is saved correctly, but the visual representation in the main design view is stale because it's reading from a different store than the one being updated.



## Testing Instructions


1. Apply the TT5 theme
1. In the Site Editor, select a new style variation, e.g., Twilight
2. Switch to View mode - verify styles match Twilight
3. Switch back to Edit mode
4. Select another style variation
5. Switch to View mode - verify styles match your new variation
6. Repeat with multiple variations to ensure each switch updates correctly
7. Also switch to a new theme and then edit a template part. 
8. The isolated editor should reflect your chose style
9. Ensure any customization to your theme show in the canvas as well.
10. Also check the templates page after switching styles. See https://github.com/WordPress/gutenberg/pull/73952#issuecomment-3645249360

## Screenshots or screencast <!-- if applicable -->

### Before

https://github.com/user-attachments/assets/6a18a110-39e5-4a01-98ff-0e161176fa9f




### After


https://github.com/user-attachments/assets/e006b818-40f3-46dc-809f-6bd3ca4f02f7

ramonjd committed:
- Import and utilize private APIs for global styles management.
- Generate global styles from merged configuration to avoid infinite loops.
- Filter out global styles from base settings and merge with generated styles.
- Update editor settings to include experimental features from global settings.

ramonjd committed:
Refactor global styles generation in editor settings

- Remove unused import of getBlockTypes.
- Simplify global styles generation by excluding block types.
- Update base settings to exclude styles and experimental features, generating them from the merged config instead.
- Ensure global settings are correctly applied to experimental features.

ramonjd committed:
comment format

ramonjd committed:
Refactor useSpecificEditorSettings to simplify base settings extraction

- Remove unused destructuring of styles and experimental features from settings.
- Streamline the process of generating base settings for the site editor.

ramonjd commented:
This also affects the template previews:


https://github.com/user-attachments/assets/d667a054-d1d8-450f-958e-45dc023dabd4

I ran out of time today to work out if there's a better, more universal fix here.

ramonjd commented on packages/edit-site/src/components/block-editor/use-site-editor-settings.js:
```diff
@@ -52,11 +60,23 @@ export function useSpecificEditorSettings() {
+		 * Exclude styles and __experimentalFeatures from base settings
+		 * since we're generating fresh ones from the merged config.
+		 */
+		const { styles, __experimentalFeatures, ...baseSettings } = settings;
```
This smells a bit. Maybe there's a better way. I ran out of time.

tellthemachines replied:
Do we need to explicitly exclude them? Won't the fresh ones just override the stale?

ramonjd replied:
Good question. Not really, I could remove it.

The intention was to mainly to make the intent clear, and:

- prevent accidentally spreading old styles if the code changes
- avoid potential issues if settings.styles has non-global styles we want to preserve

The ideal state is that we have only one source that generates global styles

ramonjd committed:
Enhance global styles integration in pattern settings

- Import private APIs for global styles management.
- Generate global styles from merged configuration to streamline settings.
- Update pattern settings to include generated styles and experimental features.
- Ensure proper memoization of global styles for performance optimization.

ramonjd commented:
> This also affects the template previews:

Meh, I just added the same fix.

ramonjd commented on packages/edit-site/src/components/page-patterns/use-pattern-settings.js:
```diff
@@ -36,16 +43,28 @@ export default function usePatternSettings() {
 		[ settingsBlockPatterns, restBlockPatterns ]
 	);
 
+	const [ globalStyles, globalSettings ] = useMemo( () => {
+		return generateGlobalStyles( mergedConfig, [], {
+			disableRootPadding: false,
+		} );
+	}, [ mergedConfig ] );
+
```
I guess we could make a hook out of all this?

ramonjd replied:
I mean to share between the two files and any others

tellthemachines replied:
If this is a temp fix pending a larger refactor then probs not worth it.

talldan commented on packages/edit-site/src/components/block-editor/use-site-editor-settings.js:
```diff
@@ -52,11 +59,23 @@ export function useSpecificEditorSettings() {
+			...baseSettings,
 			styles: [
-				...settings.styles,
+				...globalStyles,
```
I don't know the background very well, but the PR description mentions that the style settings were moved to the editor package.

I wonder why the styles need to be set here as part of site-editor settings if the `editor` package is now the source of truth. Should that happen in the editor package instead so that everything is centralized?

youknowriad replied:
I agree with the idea that ultimately, edit-site (and edit-post) should only pass "non global styles" styles to `EditorProvider` settings. and the `Editor` package should be able to read the "global styles" styles directly from the API.

I'd go even further and say even "non global styles" styles should ideally cram from a REST API but we don't really have such thing at the moment.

At least it's worth exploring the first option separately to see if it leads better results.

ramonjd replied:
💯 

My goal was to get things working as I suspect there needs to be a larger piece of work here. 

I'm not sure we can read straight from the editor store as it is - there are circular dependencies and max depth loops of death since `ExperimentalEditorProvider` updates `editorStore` on every settings change.

Here I went with `useSpecificEditorSettings()` reading only from `editSiteStore` (no `editorStore` subscription) for that reason.

But yeah, editor implementations shouldn't pass global styles to the `EditorProvider`, which means the `EditorProvider` should be defensive and exclude them from from syncing, since `GlobalStylesRenderer` handles these if I'm reading things correctly.

The tricky part is that Pattern previews use BlockPreviews, which use the `EditorProvider` directly, so GlobalStylesRenderer doesn't run.  We could create a BlockPreview wrapper in the editor package that includes GlobalStylesRenderer. I'm guessing that's not all that we'll need.

tellthemachines replied:
Don't block previews use the `ExperimentalBlockEditorProvider`? Unless I'm reading things wrong. That should be independent from any changes we make to Editor?

ramonjd replied:
> Don't block previews use the ExperimentalBlockEditorProvider? Unless I'm reading things wrong. 

I was reading it wrong 😄 Thanks for double checking. 

So maybe no wrapper required, just the addition of `GlobalStylesRenderer` here.

tellthemachines reviewed (approved):
This is working well for me! Let's get this fix in now and then we can look into something like 

> ultimately, edit-site (and edit-post) should only pass "non global styles" styles to EditorProvider settings. and the Editor package should be able to read the "global styles" styles directly from the API

without being under pressure to fix an actual regression 😅

ramonjd commented:
Thanks for testing folks. This PR fixes a pretty gnarly bug but I get that it exposes architectural debt. I agree we can fix the bug and do some planning to address the tech debt.

Essentially, edit-site should not generate global styles; GlobalStylesRenderer should be the single source of truth.

Follow ups:

- Centralize global styles generation
- Fix BlockPreview global styles

I'll see what I can knuckle out...

ramonjd committed:
Refactor global styles generation in editor and pattern settings

- Update comments to clarify the direct generation of global styles to avoid circular dependencies.
- Ensure synchronization of styles in both useSpecificEditorSettings and usePatternSettings hooks.
- Maintain consistency in global styles management across different components.

ramonjd committed:
- Remove unnecessary destructuring of styles and experimental features from settings.
